### ISSUE TITLE
Going to an event nearby should not have to restart execution during replay

### ISSUE THREAD

**Issue by sidkshatriya** [mozilla/rr#1691]
Lets say I type:

```
run 20000
```

gdb asks me if I want to restart execution from the beginning: i.e.

```
The program being debugged has been started already.
Start it from the beginning? (y or n) y
```

After waiting for a bit:
```
--------------------------------------------------
 ---> Reached target process 42895 at event 20001.
--------------------------------------------------
 ```

Now if I say `run 20010`, gdb asks me I _again_ I should start from the beginning. Why is that? Is it not possible for gdb to replay 10 more events instead of all the way from the beginning?

Or maybe there is some other syntax I need to enter in the gdb client?

My overall motivation is that I would like to use event numbers to do most of my analysis. I have some mappings from event numbers to "interesting" parts of my program ( #1688 ). Also I have a huge number of events and doing the replay from the very start will make `rr` extremely slow for me.

Now I may want to go forward in the number of events or even backward (e.g. `run 19990` ). With forward, I can imagine `rr` should be able to just replay 10 more events. In the case of going backward I would probably need to do some checkpointing perhaps.

So basically
- How to efficiently run forward a few events
- How to efficiently run backward a few events

**Comment by rocallahan** [comment]
We could easily do this with new commands. To do it with `run` we'd have to replace the `run` command entirely; `run` currently forces a symbol reload in gdb which is useful when jumping to an arbitrary event, and I don't know if you can force such a symbol reload in any way other than going through a gdb `run` command.

**Comment by sidkshatriya** [comment]
@rocallahan I'm interested in getting these new commands! Do you think this might be some functionality you might be interested in implementing? Alternatively I could take a crack at it -- I've been looking at rr source code of late but definitely need some tips to get started out on this feature :-)

I'm guessing there would be changes to classes like replaytimeline, replaytask, gdbserver and so forth...but a broad sketch of what changes might be required will help me implement this!

**Comment by rocallahan** [comment]
> Do you think this might be some functionality you might be interested in implementing?

I have other priorities right now :-).

One question is whether you want these forward/back movements to stop at breakpoints or not. Do you?

**Comment by sidkshatriya** [comment]
No its not necessary for `rr` to stop at breakpoints while going to an event. But _after_ it stops at a specific event, `rr` gdb client should be fully functional -- I should be able to set breakpoints, continue etc.

I propose we have the following syntax perhaps (though I would defer to you if you want something different)

```
go <event number> # goes to a specific event in the most efficient way based on which event we're currently at
go-forward 10 # goes 10 events forward
go-backward 10 # goes 10 events back
```

I'm happy to invest effort in implementing this to be ultimately pulled into `rr`. Some initial tips/help would be invaluable to me. I'm sure I'll probably get stuck occasionally and some help during those situations would be be awesome too :-)

**Comment by rocallahan** [comment]
I'm not sure of the best way to surface this, but in the past people have suggested having a way to stop execution at a certain event, like breakpoint-on-event. So maybe that's what we should do here too. E.g.

`first-event N` means we don't allow reverse execution to continue before event N.
`last-event N` means we don't allow forward execution to continue past event N.

Then you can implement what you want using a combination of `when`, these commands, and forward/reverse execution commands. I think this is probably the best way to go since it gives you the flexibility to trigger breakpoints/watchpoints along the way if you want.

If we take this approach, then you'd add those commands to `GdbCommands.cc` and add variables to `GdbServer` to store the first/last event number (initially 0 for `first-event` and some large dummy value for `last-event`). The first-event functionality is pretty much already implemented in `ReplayTimeline::set_reverse_execution_barrier_event`, you'd just need to make sure we never allow a value less than `first_run_event()`. The last-event functionality would be pretty easy to handle in `GdbServer`; when executing forward, deliver a SIGTRAP stop if you try to continue while the current event number is the last-event number.
